PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionContentPanelConfigurable.java

```diff
@@ -105,6 +179,137 @@ protected void createUIComponents() {
         refreshClusterListAsync();
     }
 
+    @Nullable
+    public String getLocalArtifactLibraryName() {
+        return localArtifactLibraryName;
+    }
+
+    @Nullable
+    public String getLocalArtifactPath() {
+        return localArtifactPath;
+    }
+
+    @Nullable
+    public String getLocalArtifactModuleName() {
+        return localArtifactModuleName;
+    }
+
+    public void setLocalArtifactLibraryName(@Nullable String localArtifactLibraryName) {
+        this.localArtifactLibraryName = localArtifactLibraryName;
+    }
+
+    public void setLocalArtifactPath(@Nullable String localArtifactPath) {
+        this.localArtifactPath = localArtifactPath;
+    }
+
+    public void setLocalArtifactModuleName(@Nullable String localArtifactModuleName) {
+        this.localArtifactModuleName = localArtifactModuleName;
+    }
+
+    @NotNull
+    protected String suggestModuleName(@NotNull String baseName) {
+        String candidate = baseName;
+        int idx = 1;
+        while (ModuleManager.getInstance(myProject).findModuleByName(candidate) != null) {
+            candidate = baseName + (idx++);
+        }
+        return candidate;
+    }
+
+    @Nullable
+    protected Module getOrCreateModule() {
+        // get module from module manager
+        if (getLocalArtifactModuleName() != null) {
+            Module module = ModuleManager.getInstance(myProject).findModuleByName(getLocalArtifactModuleName());
+            if (module != null) {
+                return module;
+            } else {
+                log().warn(String.format("module %s is created but can't find the module", getLocalArtifactModuleName()));
+            }
+        }
+        // create new module
+        String moduleName = suggestModuleName(LOCAL_ARTIFACT_MODULE_NAME);
+        File moduleFile = new File(myProject.getBasePath(), String.format("%s/%s.iml", moduleName, moduleName));
+        Module module = ModuleManager.getInstance(myProject).newModule(moduleFile.getPath(), ModuleTypeId.JAVA_MODULE);
+        setLocalArtifactModuleName(module.getName());
+        return module;
+    }
+
+    public void addJarToProjectLibrary(@Nullable VirtualFile jarFile) throws IOException {
+        if (jarFile == null) {
+            return;
+        }
+
+        final List<OrderRoot> myRoots = RootDetectionUtil.detectRoots(
```

Comment:
Could you add some comments for the following codes?

Reply:
I find the following codes add library in PROJECT level, so I refactor the code to add library in MODULE level

---
In reply to: [209543878](https://github.com/Microsoft/azure-tools-for-java/pull/1848#discussion_r209543878) [](ancestors = 209543878)